src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,1520 @@
+  Internal utility routines
+  ===============================================================*/
+
+/*---------------------------------------------------------------
+  lsrkStep_AccessARKODEStepMem:
+
+  Shortcut routine to unpack both ark_mem and step_mem structures
+  from void* pointer.  If either is missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessARKODEStepMem(void* arkode_mem, const char* fname,
+                                 ARKodeMem* ark_mem, ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeMem structure */
+  if (arkode_mem == NULL)
+  {
+    arkProcessError(NULL, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_ARK_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *ark_mem = (ARKodeMem)arkode_mem;
+
+  /* access ARKodeLSRKStepMem structure */
+  if ((*ark_mem)->step_mem == NULL)
+  {
+    arkProcessError(*ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *step_mem = (ARKodeLSRKStepMem)(*ark_mem)->step_mem;
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_AccessStepMem:
+
+  Shortcut routine to unpack the step_mem structure from
+  ark_mem.  If missing it returns ARK_MEM_NULL.
+  ---------------------------------------------------------------*/
+int lsrkStep_AccessStepMem(ARKodeMem ark_mem, const char* fname,
+                           ARKodeLSRKStepMem* step_mem)
+{
+  /* access ARKodeLSRKStepMem structure */
+  if (ark_mem->step_mem == NULL)
+  {
+    arkProcessError(ark_mem, ARK_MEM_NULL, __LINE__, fname, __FILE__,
+                    MSG_LSRKSTEP_NO_MEM);
+    return (ARK_MEM_NULL);
+  }
+  *step_mem = (ARKodeLSRKStepMem)ark_mem->step_mem;
+  return (ARK_SUCCESS);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_CheckNVector:
+
+  This routine checks if all required vector operations are
+  present.  If any of them is missing it returns SUNFALSE.
+  ---------------------------------------------------------------*/
+sunbooleantype lsrkStep_CheckNVector(N_Vector tmpl)
+{
+  if ((tmpl->ops->nvclone == NULL) || (tmpl->ops->nvdestroy == NULL) ||
+      (tmpl->ops->nvlinearsum == NULL) || (tmpl->ops->nvconst == NULL) ||
+      (tmpl->ops->nvscale == NULL) || (tmpl->ops->nvwrmsnorm == NULL) ||
+      (tmpl->ops->nvspace == NULL))
+  {
+    return (SUNFALSE);
+  }
+  return (SUNTRUE);
+}
+
+/*---------------------------------------------------------------
+  lsrkStep_ComputeNewDomEig:
+
+  This routine computes new dom_eig and returns SUN_SUCCESS.
+  ---------------------------------------------------------------*/
+
+int lsrkStep_ComputeNewDomEig(ARKodeMem ark_mem, ARKodeLSRKStepMem step_mem)
+{
+  int retval = SUN_SUCCESS;
+
+  if ((step_mem->is_ext_dom_eig))
+  {
+    retval = step_mem->extDomEig(ark_mem->tn, ark_mem->ycur, ark_mem->fn,
```

Comment:
Should we track the number of calls to the user-provided dominant eigenvalue function?

Reply:
We used to do it after `lsrkStep_ComputeNewDomEig` call. It is now right after the function call.